google-home-community.groovy

```diff
@@ -2886,9 +2894,8 @@ private attributesForTrait_Brightness(deviceTrait) {
 @SuppressWarnings(['UnusedPrivateMethod', 'UnusedPrivateMethodParameter'])
 private attributesForTrait_CameraStream(deviceTrait) {
     return [
-        cameraStreamSupportedProtocols: deviceTrait.cameraStreamSupportedProtocols,
+        cameraStreamSupportedProtocols: ["progressive_mp4", "hls", "dash", "smooth_stream"],
```

Comment:
Should we get this list from a device attribute instead of assuming all devices support all streaming protocols?

Reply:
Issue I'm seeing is we don't know what each device's supported protocol is.  I suppose maybe I could query all of them, and then generate a list to pass back?  If the display device supports Chromecast, then it should support all of these.

Reply:
I'm pretty sure this attribute tells Google what protocols the _source_ supports, not the display.  I was thinking something like what Hubitat does with `supportedThermostatModes`.  Add an attribute to the driver that can be used to report supported streaming protocols to the app.  For the generic driver the user would have to set it, but for any specific devices it could be hard-coded into the driver.

Reply:
Issue is what the app supports not the driver.  It's only called at the start of the sync process, and that result is then loaded into all the camera devices reported back to Google.   The only time the driver level support is pulled is during the execute via deviceInfo.device.currentValue(deviceInfo.deviceType.traits.CameraStream.cameraStreamProtocolAttribute.  

It used to be like supportedThermostatModes where you could pick the supported protocols, but that prevented the user from linking more than one.  I can add it back to the app, but then it becomes confusing for the user since they need to configure it in the app as well as the driver level.  Thoughts?

Reply:
> Issue is what the app supports not the driver. It's only called at the start of the sync process, and that result is then loaded into all the camera devices reported back to Google. The only time the driver level support is pulled is during the execute via deviceInfo.device.currentValue(deviceInfo.deviceType.traits.CameraStream.cameraStreamProtocolAttribute.
> 
> It used to be like supportedThermostatModes where you could pick the supported protocols, but that prevented the user from linking more than one. I can add it back to the app, but then it becomes confusing for the user since they need to configure it in the app as well as the driver level. Thoughts?

I re-looked at this, but still running back to this issue.  The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?

Reply:
> I re-looked at this, but still running back to this issue. The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?

I see what you mean now.  I think it would make sense to move the loop line 2910 down into the following loop and pass both the device and the trait into the `attributesForTrait_*` methods.  That way the initial SYNC would have access to the devices themselves to pull things like supported protocols.

The other option would be to move it back to being a trait level setting and make users define multiple device types for devices that support different protocols.  That would also make sense to me, since all instances of one type of camera will always support the same protocols.

Reply:
Let me take a look again with those suggestions.  Thanks!

Reply:
> > I re-looked at this, but still running back to this issue. The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?
> 
> I see what you mean now. I think it would make sense to move the loop line 2910 down into the following loop and pass both the device and the trait into the `attributesForTrait_*` methods. That way the initial SYNC would have access to the devices themselves to pull things like supported protocols.
> 
> The other option would be to move it back to being a trait level setting and make users define multiple device types for devices that support different protocols. That would also make sense to me, since all instances of one type of camera will always support the same protocols.

Struggling a bit on this one.  Won't this change break all the existing code actions since the traits are global, and not device specific?  I understand your concern with the protocols, but from the testing I've done, if Google asks for something the driver cannot handle, it handles it gracefully on their end.

Reply:
The SYNC request handler already does device-specific stuff to get default rooms, there's no reason it can't do other device-specific lookups.  I'm basically just suggesting to change it from this:

```
    (deviceTypes() + [modeSceneDeviceType()]).each { deviceType ->
        def traits = deviceType.traits.collect { traitType, deviceTrait ->
            "action.devices.traits.${traitType}"
        }
        def attributes = [:]
        deviceType.traits.each { traitType, deviceTrait ->
            attributes += "attributesForTrait_${traitType}"(deviceTrait)
        }
        deviceType.devices.each { device ->
            //...
        }
```

To something like this:

```
    (deviceTypes() + [modeSceneDeviceType()]).each { deviceType ->
        def traits = deviceType.traits.collect { traitType, deviceTrait ->
            "action.devices.traits.${traitType}"
        }
        deviceType.devices.each { device ->
            def attributes = [:]
            deviceType.traits.each { traitType, deviceTrait ->
                attributes += "attributesForTrait_${traitType}"(deviceTrait, device)
            }

            // ...
        }
```

The `attributesForTrait_*` methods would need to accept the second parameter, but most could just ignore it and always return the same thing for all devices.